### ISSUE TITLE
Don't sum over an index for terms that lack it

### ISSUE THREAD

**Issue by sethaxen** [ahwillia/Einsum.jl#39]
Suppose I have the following case

```julia
julia> using Einsum;
julia> a = [1, 2];
julia> b = [1 2; 3 4];
julia> c = a+b*a  # what I would like to compute
2-element Array{Int64,1}:
  6
 13
julia> @einsum c[i] := a[i] + b[i,j] * a[j]  # this doesn't work
2-element Array{Int64,1}:
  7
 15
julia> @einsum c[i] := b[i,j] * a[j];  # breaking up the problem works
julia> @einsum c[i] += a[i]
2-element Array{Int64,1}:
  6
 13
```
It seems terms that don't include an index are still summed over that index. From my understanding of Einstein notation, this isn't supposed to happen. Is there a way to avoid the hack above? For reference, `TensorOperations.jl` handles the sum as expected.

```julia
julia> using TensorOperations;
julia> @tensor c[i] := a[i] + b[i,j] * a[j]
2-element Array{Int64,1}:
  6
 13
```

**Comment by mcabbott** [comment]
I agree that's the standard meaning of Einstein notation. But it would be far from simple to make this package do that, and if you wanted it to be efficient, you'd end up re-writing TensorOperations.jl. 

Despite the name, what this package seems good at doing is other sums, or broadcast-like operations, especially with functions other than `*` and `+` on the right. Like `m(pq, p, q)` in the readme. (It is, sadly, quite often faster than TensorCast.jl on these things.) For these the rule of summing over all free indices is unambiguous, but again not very Einstein. 

The package ArrayMeta.jl also obeyed the same rule: 
```
julia> using ArrayMeta # Julia 0.6, or 0.7 with the right PR

julia> @arrayop c[i] := a[i] + b[i,j] * a[j]
2-element Array{Int64,1}:
  7
 15

julia> c = sum(@. a + b * a'; dims=2)
2×1 Array{Int64,2}:
  7
 15
```
